File: lib/executor.rb
require 'json'

class Executor
  def initialize(data, part1_proc, part2_proc)
    @data = data
    @part1_proc = part1_proc
    @part2_proc = part2_proc
  end

  def run(args)
    args[-1] == "--time" ? time : solve
  end

  def solve
    puts @part1_proc.call(@data.dup)
    puts @part2_proc.call(@data.dup)
  end

  private

  def time
    puts JSON.generate({
      part1: time_proc(@part1_proc),
      part2: time_proc(@part2_proc)
    })
  end

  def continue_timing?(iterations, duration)
    if iterations < 100
      return duration < 30
    else
      return duration < 0.1
    end
  end

  def time_proc(fn)
    i = 0
    start_time = Time.now.to_f
    while continue_timing?(i, Time.now.to_f - start_time)
      fn.call(@data.dup)
      i += 1
    end
    {duration: ((Time.now.to_f - start_time) * 1_000_000).round, iterations: i}
  end
end

File: lib/python/lib/executor.py
import json

from datetime import datetime


class Executor(object):
    def __init__(self, data, part1_fn, part2_fn):
        self.data = data
        self.part1_fn = part1_fn
        self.part2_fn = part2_fn

    def __call__(self, argv):
        if argv[-1] == "--time":
            self._time()
        else:
            self._solve()

    def _solve(self):
        print(self.part1_fn([v for v in self.data]))
        print(self.part2_fn([v for v in self.data]))

    def _time(self):
        print(
            json.dumps(
                {
                    "part1": self._time_fn(self.part1_fn),
                    "part2": self._time_fn(self.part2_fn),
                }
            )
        )

    def _time_fn(self, fn):
        def duration_us(duration):
            return duration.seconds * 1000 * 1000 + duration.microseconds

        def continue_timing(iterations, duration):
            if iterations < 100:
                return duration < 30000000
            else:
                return duration < 100000

        i = 0
        running_time = 0
        while continue_timing(i, running_time):
            data = [v for v in self.data]
            start_time = datetime.now()
            fn(data)
            running_time += duration_us(datetime.now() - start_time)
            i += 1
        return {"duration": running_time, "iterations": i}
